File: DraftRightWindows/DraftRightWindows/Models/Tone.cs
namespace DraftRightWindows.Models;

public enum Tone
{
    Simple,
    Natural,
    Polished,
    Concise,
    Technical,
    Claude,
    GrammarCheck,
    Translate
}

public static class ToneExtensions
{
    public static string ApiValue(this Tone tone) => tone switch
    {
        Tone.Simple => "simple",
        Tone.Natural => "natural",
        Tone.Polished => "polished",
        Tone.Concise => "concise",
        Tone.Technical => "technical",
        Tone.Claude => "claude",
        Tone.GrammarCheck => "grammar_check",
        Tone.Translate => "translate",
        _ => "polished"
    };

    /// <summary>Parse a backend API value (e.g. "polished", "grammar_check")
    /// back into a <see cref="Tone"/>, or null if it doesn't match a known
    /// tone. Inverse of <see cref="ApiValue"/>.</summary>
    public static Tone? FromApiValue(string? apiValue) => apiValue switch
    {
        "simple" => Tone.Simple,
        "natural" => Tone.Natural,
        "polished" => Tone.Polished,
        "concise" => Tone.Concise,
        "technical" => Tone.Technical,
        "claude" => Tone.Claude,
        "grammar_check" => Tone.GrammarCheck,
        "translate" => Tone.Translate,
        _ => null
    };

    /// <summary>
    /// True when the rewrite result depends on the configured target language,
    /// so callers must send it to the backend and fold it into the cache key.
    /// Mirrors Linux's <c>Tone.uses_target_language</c>. Kept on the enum so a
    /// future language-aware tone is one line here rather than another
    /// `tone == Tone.Translate` literal scattered through the UI (Rule #1).
    /// </summary>
    public static bool UsesTargetLanguage(this Tone tone) => tone == Tone.Translate;

    /// <summary>
    /// True when the tone's result is plain text that can be cached as a string.
    /// False for tones returning a structured payload — caching those as a
    /// string would silently drop the structure (Grammar Check's issues list).
    ///
    /// Must gate the cache read AND the cache write. Reading with one condition
    /// and writing with another produces entries that can never be hit but
    /// still consume the bounded cache — the same asymmetry that made Linux's
    /// translate results permanently un-cacheable (#108).
    /// </summary>
    public static bool ProducesCacheableText(this Tone tone) => tone != Tone.GrammarCheck;

    public static string DisplayName(this Tone tone) => tone switch
    {
        Tone.Simple => "Simple",
        Tone.Natural => "Natural",
        Tone.Polished => "Polished",
        Tone.Concise => "Concise",
        Tone.Technical => "Technical",
        Tone.Claude => "Claude Style",
        Tone.GrammarCheck => "Grammar Check",
        Tone.Translate => "Translate",
        _ => "Polished"
    };

    // Use BMP-only glyphs so Segoe UI renders them directly without falling
    // back to Segoe UI Emoji (which often shows as tofu boxes in WinForms).
    public static string Icon(this Tone tone) => tone switch
    {
        Tone.Simple       => "\u270e",  // \u270e pencil
        Tone.Natural      => "\u275d",  // \u275d heavy double turned comma quotation
        Tone.Polished     => "\u2728",  // \u2728 sparkles (renders fine in Segoe UI)
        Tone.Concise      => "\u2296",  // \u2296 circled minus
        Tone.Technical    => "\u2699",  // \u2699 gear
        Tone.Claude       => "\u24b8",  // \u24b8 circled C
        Tone.GrammarCheck => "\u2713",  // \u2713 check mark
        Tone.Translate    => "\u21c4",  // \u21c4 rightwards arrow over leftwards
        _ => "\u2728"
    };
}

File: DraftRightLinux/draftright/models/tone.py
from enum import Enum


class Tone(Enum):
    SIMPLE = ("simple", "Simple", "\u270e", "Easy language, short sentences")
    NATURAL = ("natural", "Natural", "\U0001f4ac", "Conversational, flows smoothly")
    POLISHED = ("polished", "Polished", "\u2728", "Professional, refined")
    CONCISE = ("concise", "Concise", "\u2296", "Removes redundancy")
    TECHNICAL = ("technical", "Technical", "\U0001f527", "Precise, documentation style")
    CLAUDE = ("claude", "Claude Style", "\U0001f916", "Clear, thoughtful, well-structured")
    GRAMMAR_CHECK = ("grammar_check", "Grammar Check", "\u2713", "Check spelling, grammar, style")
    TRANSLATE = ("translate", "Translate", "\U0001f310", "Translate to another language")

    def __init__(self, api_value, display_name, icon, description):
        self.api_value = api_value
        self.display_name = display_name
        self.icon = icon
        self.description = description

    @property
    def uses_target_language(self) -> bool:
        """True when the result depends on the configured translate language.

        Drives both what is sent to /rewrite and what goes into the cache key
        — without it, switching language would serve the previous language's
        translation from cache.
        """
        return self is Tone.TRANSLATE

    @property
    def returns_grammar_analysis(self) -> bool:
        """True when /rewrite answers with a ``grammar`` object, not text.

        Grammar Check is the one tone whose response has no ``rewritten_text``
        — parsing it as one raised "/rewrite response has no rewritten_text"
        and the tone did nothing but show an error (#107).
        """
        return self is Tone.GRAMMAR_CHECK

    @property
    def produces_replacement_text(self) -> bool:
        """True when the result can be pasted straight over the selection.

        One-Click has no UI to review anything, so it can only offer tones that
        yield replacement text — a grammar analysis has nothing to inject.
        """
        return not self.returns_grammar_analysis

    @classmethod
    def from_api_value(cls, value: str) -> "Tone | None":
        for tone in cls:
            if tone.api_value == value:
                return tone
        return None
